cmd/auth-provider-gcp/app/getcredentials_test.go

```diff
@@ -0,0 +1,92 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestValidateAuthFlow(t *testing.T) {
+	type FlagResult struct {
+		Flow  string
+		Error error
+	}
+	tests := []FlagResult{
+		{Flow: gcrAuthFlow, Error: nil},
+		{Flow: dockerConfigAuthFlow, Error: nil},
+		{Flow: dockerConfigURLAuthFlow, Error: nil},
+		{Flow: "bad-flow", Error: &AuthFlowFlagError{flagValue: "bad-flow"}},
+		{Flow: "", Error: &AuthFlowFlagError{flagValue: ""}},
+		{Flow: "Gcrauthflow", Error: &AuthFlowFlagError{flagValue: "Gcrauthflow"}},
+	}
+	for _, tc := range tests {
+		err := validateFlags(&CredentialOptions{AuthFlow: tc.Flow})
+		if err != nil && tc.Error == nil {
+			t.Errorf("with flow %q unexpected error %q", tc.Flow, err)
+		}
+		if err == nil && tc.Error != nil {
+			t.Errorf("with flow %q did not get expected error %q", tc.Flow, err)
+		}
+		if err != nil && tc.Error != nil {
+			if reflect.TypeOf(err) != reflect.TypeOf(tc.Error) {
+				t.Errorf("with flow %q got unexpected error type %q (expected %q)", tc.Flow, reflect.TypeOf(err), reflect.TypeOf(tc.Error))
+			}
+		}
+	}
+}
+
+func TestProviderFromFlow(t *testing.T) {
+	type ProviderResult struct {
+		Flow  string
+		Type  string
+		Error error
+	}
+	tests := []ProviderResult{
+		{Flow: gcrAuthFlow, Type: "ContainerRegistryProvider", Error: nil},
+		{Flow: dockerConfigAuthFlow, Type: "DockerConfigKeyProvider", Error: nil},
+		{Flow: dockerConfigURLAuthFlow, Type: "DockerConfigURLKeyProvider", Error: nil},
+		{Flow: "bad-flow", Type: "", Error: &AuthFlowTypeError{requestedFlow: "bad-flow"}},
+		{Flow: "", Type: "", Error: &AuthFlowTypeError{requestedFlow: ""}},
+	}
+	for _, tc := range tests {
+		provider, err := providerFromFlow(tc.Flow)
+		if err != nil && tc.Error == nil {
+			t.Errorf("with flow %q unexpected error %q", tc.Flow, err)
+		}
+		if err == nil && tc.Error != nil {
+			t.Errorf("with flow %q did not get expected error %q", tc.Flow, err)
+		}
+		if err != nil && tc.Error != nil {
+			if reflect.TypeOf(err) != reflect.TypeOf(tc.Error) {
+				t.Errorf("with flow %q got unexpected error type %q (expected %q)", tc.Flow, reflect.TypeOf(err), reflect.TypeOf(tc.Error))
+			}
+		}
+		if tc.Type == "" && provider != nil {
+			t.Errorf("with flow %q got unexpectedly non-nil provider %q", provider)
+		}
+		// The nil check is meant for test cases where provider is nil on purpose,
+		// i,e, for error cases - any errors will get tested and caught above, so in
+		// those cases, we don't need to run additional checks
+		if provider != nil {
```

Comment:
Insteawd of `if provider != nil`, use `t.Fatalf` on the above failures. This way, execution only gets to this point if the provider should exist.

This has the advantage of (1) being easier to read, and (2) properly covering the `provider == nil && err == nil` case.

For this to work properly, also need to do something like:

```go
for _, tc := range test {
  tc.Run(func(tc.Name, t *testing.T) {
    
  })
}
```

This is needed since, without the nested `Run()`, the `Fatalf` would terminate the for loop.

Recommend also finding a way to name the tests in the 1st argument to run in some way that is clear, either by adding an explicit "Name" field to the test case or by concatenating existing test case fields in a way that is sufficiently clear. 

Applies to both added tests in this file.

Reply:
Makes sense - done!